openmmmcmc/states.py

```diff
@@ -0,0 +1,1903 @@
+        particles.
+
+        Raises
+        ------
+        SamplerStateError
+            If set to an array with a number of particles different
+            than n_particles.
+
+        """
+        return self._velocities
+
+    @velocities.setter
+    def velocities(self, value):
+        if value is not None and self.n_particles != len(value):
+            raise SamplerStateError(SamplerStateError.INCONSISTENT_VELOCITIES)
+        self._velocities = value
+        self._cached_velocities_in_md_units = None  # Invalidate cache.
+
+    @property
+    def total_energy(self):
+        """The sum of potential and kinetic energy (read-only)."""
+        if self.potential_energy is None or self.kinetic_energy is None:
+            return None
+        return self.potential_energy + self.kinetic_energy
+
+    @property
+    def volume(self):
+        """The volume of the box (read-only)"""
+        return _box_vectors_volume(self.box_vectors)
+
+    @property
+    def n_particles(self):
+        """Number of particles (read-only)."""
+        return len(self.positions)
+
+    def is_context_compatible(self, context):
+        """Check compatibility of the given context.
+
+        The context is compatible if this SamplerState can be applied
+        through apply_to_context.
+
+        Parameters
+        ----------
+        context : simtk.openmm.Context
+            The context to test.
+
+        Returns
+        -------
+        is_compatible : bool
+            True if this SamplerState can be applied to context.
+
+        See Also
+        --------
+        SamplerState.apply_to_context
+
+        """
+        is_compatible = self.n_particles == context.getSystem().getNumParticles()
+        return is_compatible
+
+    def update_from_context(self, context):
+        """Read the state from the given context.
+
+        The context must be compatible. Use SamplerState.from_context
+        if you want to build a new sampler state from an incompatible.
+
+        Parameters
+        ----------
+        context : simtk.openmm.Context
+            The context to read.
+
+        Raises
+        ------
+        SamplerStateError
+            If the given context is not compatible.
+
+        """
+        self._read_context_state(context, check_consistency=True)
+
+    def apply_to_context(self, context):
+        """Set the context state.
+
+        If velocities and box vectors have not been specified in the
+        constructor, they are not set.
+
+        Parameters
+        ----------
+        context : simtk.openmm.Context
+            The context to set.
+
+        """
+        context.setPositions(self._positions_in_md_units)
+        if self._velocities is not None:
+            context.setVelocities(self._velocities_in_md_units)
+        if self.box_vectors is not None:
+            context.setPeriodicBoxVectors(*self.box_vectors)
+
+    def __getitem__(self, item):
```

Comment:
There's an example at the end of the class documentation, but I'll duplicate the docs here too! I report it here:
```
    It is possible to slice a sampler state to obtain positions and
    particles of a subset of atoms

    >>> sliced_sampler_state = sampler_state[:10]
    >>> sliced_sampler_state.n_particles
    10
```
It basically return a new `SamplerState` with only positions/velocities of the sliced particles.